### PR TITLE
Backport Kafka access log to v0.12

### PR DESCRIPTION
**Pull request by manalibhutiyani**
Issue #1937 Backport Kafka access log to v0.12
Fixes: #1937

```release-note
Kafka access log support
```

### CODE REVIEW

File: pkg/policy/api/selector.go [hunk 1]
```diff
@@ -21,18 +21,13 @@ import (
 	"github.com/cilium/cilium/pkg/labels"
 	"github.com/cilium/cilium/pkg/logfields"
 
+	"fmt"
```
Comment (aanm) [inline pkg/policy/api/selector.go]: It's seems at least this commit was created for this branch, you should merge this commit first on master and then back port it to `v0.12`

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/policy/api/selector.go (hunk 1 above)
Comment (manalibhutiyani) [reply pkg/policy/api/selector.go]: I cannot merge this into master. The new logging changes done by Ray in master are not to be back-ported and this was a way to get around the merge conflicts that created.

Comment (aanm) [reply pkg/policy/api/selector.go]: My comment was for the whole commit, I was also referring on the method's descriptions changes bellow.

**Review by tgraf — APPROVED** [review]

**Comment by manalibhutiyani** [comment]
@raybejjani @jrajahalme I need an approval from proxy side to merge this..please take a look.

File: pkg/proxy/accesslog/log.go [hunk 2]
```diff
@@ -94,42 +119,20 @@ func logString(outStr string, retry bool) {
...
-		FieldHeader:   l.Request.Header,
-		FieldFilePath: logPath,
-	}).Debug("Logging L7 flow record")
-
 	if logBuf == nil {
-		log.WithFields(log.Fields{
-			FieldFilePath: logPath,
-		}).Debug("Skipping writing to access log (write buffer nil)")
+		log.WithField(FieldFilePath,
+			logPath).Debug("Skipping writing to access log (write buffer nil)")
```
Comment (raybejjani) [inline pkg/proxy/accesslog/log.go]: aanm has shown me a nifty trick. It reads nicer but feels completely weird. We can split between the chained functions, if you prefer:
```golang
log.WithField(FieldFilePath, logPath).
    Debug("Skipping writing to access log (write buffer nil)")
```

File: pkg/proxy/oxyproxy.go [hunk 3]
```diff
@@ -344,83 +184,61 @@ func createOxyRedirect(l4 *policy.L4Filter, id string, source ProxySource, to ui
...
 				redir.mutex.Unlock()
-				accesslog.Log(record, accesslog.TypeRequest, accesslog.VerdictDenied, http.StatusForbidden)
+				record.log(accesslog.TypeRequest, accesslog.VerdictDenied,
+					http.StatusForbidden, "")
 				return
 			}
 			ar := rule.(string)
-			log.Debugf("Allowing request based on rule %+v", ar)
-			record.Info = fmt.Sprintf("rule: %+v", ar)
+			log.WithField(logfields.Object, fmt.Sprintf("%+v", "Allowing request based on rule"))
```
Comment (raybejjani) [inline pkg/proxy/oxyproxy.go]: This is missing the Debug call
```golang
log.WithField(logfields.Object, fmt.Sprintf("%+v", ar).Debug("Allowing request based on rule")
```
We can also backport the logfields.Repr function but it basically just does the fmt.Sprintf call

File: pkg/policy/api/selector.go [hunk 4]
```diff
@@ -157,7 +155,8 @@ func (n *EndpointSelector) Matches(lblsToMatch k8sLbls.Labels) bool {
 		// FIXME: Omit this error or through it to the caller?
 		// We are doing the verification in the ParseEndpointSelector but
 		// don't make sure the user can modify the current labels.
-		log.Errorf("unable the match selector %+v in selector: %s", n, err)
+		log.WithError(err).WithField(logfields.EndpointLabelSelector,
+			fmt.Sprintf("%+v", "unable to match label selector in selector"))
```
Comment (raybejjani) [inline pkg/policy/api/selector.go]: This is missing the `Error` call
```golang
log.WithError(err).WithField(logfields.EndpointLabelSelector, fmt.Sprintf("%+v",n).
    Error("unable to match label selector in selector")
```

File: pkg/proxy/kafka.go [hunk 5]
```diff
@@ -258,24 +351,53 @@ func handleConnection(pair *connectionPair, c *proxyConnection, handler kafkaMes
...
+func (k *kafkaRedirect) handleResponseConnection(pair *connectionPair,
+	record *kafkaLogRecord) {
 	log.WithFields(log.Fields{
 		"from": pair.tx,
 		"to":   pair.rx,
 	}).Debug("Proxying response Kafka connection")
 
-	handleConnection(pair, pair.tx, func(pair *connectionPair, req *kafka.RequestMessage) {
-		pair.rx.Enqueue(req.GetRaw())
+	handleResponse(pair, pair.tx, record, func(pair *connectionPair,
```
Comment (raybejjani) [inline pkg/proxy/kafka.go]: It might be more clear if we split the line before the func

**Review by raybejjani — CHANGES_REQUESTED** [review]
My comments are mostly about formatting or non-printing log messages. Feel free to override this if you prefer to fix those in a separate PR.

File: pkg/policy/api/selector.go (hunk 4 above)
Comment (tgraf) [reply pkg/policy/api/selector.go]: fixed

File: pkg/proxy/accesslog/log.go (hunk 2 above)
Comment (tgraf) [reply pkg/proxy/accesslog/log.go]: fixed

File: pkg/proxy/oxyproxy.go (hunk 3 above)
Comment (tgraf) [reply pkg/proxy/oxyproxy.go]: fixed

File: pkg/proxy/kafka.go (hunk 5 above)
Comment (tgraf) [reply pkg/proxy/kafka.go]: fixed

**Review by aanm — APPROVED** [review]

**Review by raybejjani — APPROVED** [review]

**Pull request merged**